### PR TITLE
feat(sync): bridge SAF-granted URI to aw-sync output (Phase 2)

### LINKED ISSUE #1357: aw-sync: Android sync UX needs a complete pass (setup flow, default-off, storage access)
## Background

Filed out of discussion in [ActivityWatch/aw-server-rust#630](https://github.com/ActivityWatch/aw-server-rust/issues/630), which traced Android native OOM crashes to an unbounded event fetch in `sync_one()`.

In that thread @ErikBjare asked the right follow-up: **what does the user actually do to get their Android and desktop data synced?** The answer is: there is no user-facing setup flow, and the underlying mechanism is blocked by Android storage restrictions.

## Current state (Android sync, as implemented)

Android sync dir is resolved as `Android/data/net.activitywatch.android/files/sync/` (set via `AW_SYNC_DIR` env var in `SyncInterface.kt`). The sync DB structure mirrors desktop: `{hostname}/{device_id}/*.db`.

- **Push** (`push_with_hostname`): Android writes its own events into `sync_dir/{device_name}/{device_id}/*.db`. This works in isolation (modulo the OOM bug in #630).
- **Pull** (`pull_all`): Android scans `sync_dir` for `{hostname}/{device_id}/*.db` files from OTHER devices. **There is no mechanism to get those files onto the phone.** On desktop, users share `~/ActivityWatchSync` via Syncthing/Dropbox/etc. On Android, the app's data dir (`Android/data/...`) is scoped storage — not accessible to other apps since Android 11+, so Syncthing cannot write into it, making the typical desktop sync flow non-replicable on mobile.

In practice: `pull_all()` silently no-ops on a typical Android device (no remote db files found). Push works, but the data stays trapped on the phone.

## Problems to fix

1. **Sync disabled by default on Android** — automatic sync fires every 15 min (via `SyncScheduler`), potentially crashing (OOM) or consuming battery/I/O for work that is currently pointless. It should be off by default, with a setting to enable it.

2. **Android sync directory is not accessible via Syncthing/standard apps** — the app data dir is scoped to the app. For Android sync to work at all, either:
   - The app needs to use a user-accessible location (e.g., a user-picked directory via Storage Access Framework / `ACTION_OPEN_DOCUMENT_TREE`)
   - Or provide its own sync transport (Bluetooth, local Wi-Fi, explicit export/share flow)

3. **No setup UI** — no settings screen, no documentation, no in-app explanation of how sync works or what the user needs to do.

4. **SyncScheduler runs unconditionally** — no check for whether sync is configured, no user-accessible enable/disable toggle.

## Suggested next steps (not prescriptive — up to maintainers)

- Add a settings toggle to enable/disable automatic sync (default: off)
- Consider Storage Access Framework for Android sync dir (user picks a directory accessible to both the app and Syncthing/etc.)
- Add a sync status/setup screen explaining the current sync dir path and what users need to do
- Document the intended mobile sync flow somewhere visible

## Related

- ActivityWatch/aw-server-rust#630 — OOM crash from unbounded fetch (separate fix, but the same sync path)
- `aw-android/mobile/src/main/java/net/activitywatch/android/SyncInterface.kt` — sync dir setup
- `aw-android/mobile/src/main/java/net/activitywatch/android/SyncScheduler.kt` — 15-min auto-sync timer

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Phase 2 of ActivityWatch/activitywatch#1357. Phase 1 (#204) added the UI and SAF permission infrastructure. This closes the loop by actually copying sync files to the user-chosen directory.

**Problem**: `aw-sync` writes to `getExternalFilesDir()/sync/` which has been inaccessible to Syncthing and other file-sync tools since Android 11 (scoped storage). Phase 1 let users pick an SAF directory and stored the grant persistently — but nothing yet wrote files there.

**Solution**: After each successful sync operation, `copySyncFilesToSafDir()` mirrors every file from the internal sync dir to the SAF-granted location using `DocumentFile` and `ContentResolver`.

Key design decisions:
- **File-copy approach** (not DocumentFile I/O rewrite): keeps the native `aw-sync` library untouched; simplest path that makes files accessible where external tools can reach them
- **Background thread only**: copy runs on the existing executor thread after `syncFn()` completes — never on the main thread
- **Non-fatal errors**: `IOException` and `SecurityException` are logged and counted but never propagate; a copy failure must not fail the sync itself
- **Explicit `documentfile` dep**: was transitive via `appcompat`, now declared (`1.1.0`)

## Changes

- `SyncInterface.kt`: add `copySyncFilesToSafDir()` + call from `performSyncAsync` on each successful sync
- `build.gradle`: explicit `androidx.documentfile:documentfile:1.1.0` dependency

## Test plan

- [ ] Build debug APK and install on a physical Android 11+ device
- [ ] Open Sync Settings → configure a directory accessible to Syncthing (or Files app)
- [ ] Enable sync toggle
- [ ] Trigger a manual sync (or wait for scheduled sync)
- [ ] Verify sync files appear in the chosen directory (e.g. via Files app or Syncthing)
- [ ] Verify that choosing a new SAF directory on next sync causes files to appear there
- [ ] Verify that no SAF dir configured → no copy attempt (internal sync dir only)

Note: SAF integration is exercised via instrumented tests on-device; `logcat` tag `SyncInterface` shows `SAF mirror: copied=N skipped=M → content://...` on each sync when a directory is configured.

Closes ActivityWatch/activitywatch#1357 (Phase 2 of 2)

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 1]
```diff
@@ -139,6 +146,55 @@ class SyncInterface(context: Context) {
...
+     * file-sync tools on Android 11+. This method copies every regular file from [syncDir]
+     * to the SAF-granted tree URI so that external sync tools can reach the data.
+     *
+     * Errors are logged but do not propagate — a copy failure must never fail the sync itself.
+     */
+    private fun copySyncFilesToSafDir() {
+        val uriStr = AWPreferences(appContext).getSyncDirUri() ?: return
+        val safUri = Uri.parse(uriStr)
+        val safDir = DocumentFile.fromTreeUri(appContext, safUri)
+        if (safDir == null || !safDir.isDirectory) {
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:229]: Fixed in 6b6c4fb: the `copySyncFilesToSafDir()` call is now wrapped in a `try { ... } catch (e: Exception)` block in `performSyncAsync`, so any `SecurityException` thrown by `isDirectory` or `fromTreeUri` is caught and logged as non-fatal. It can never propagate to the outer `performSyncAsync` catch and misreport the sync as failed.

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 2]
```diff
@@ -139,6 +146,55 @@ class SyncInterface(context: Context) {
...
+                    ?: safDir.createFile("application/octet-stream", file.name)
+                if (dest == null) {
+                    Log.w(TAG, "Could not create SAF file for ${file.name}")
+                    skipped++
+                    continue
+                }
+                appContext.contentResolver.openOutputStream(dest.uri, "wt")?.use { out ->
+                    FileInputStream(file).use { inp -> inp.copyTo(out) }
+                }
+                copied++
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt]: Fixed in 6b6c4fb: the `?.use {}` nullable shortcut was replaced with an explicit null check — if `openOutputStream` returns null, we log a warning, increment `skipped`, and `continue`. The `copied++` increment only happens after a successful `out.use { ... }` block.

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 3]
```diff
@@ -123,7 +127,10 @@ class SyncInterface(context: Context) {
                 } else {
                     json.getString("error")
                 }
-                
+
+                // Copy sync files to SAF directory on the background thread — IO must not run on main.
+                if (success) copySyncFilesToSafDir()
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt]: Fixed in 6b6c4fb: the callback is now posted to the main thread (`handler.post { callback(success, message) }`) *before* the SAF mirror runs. The `AlarmReceiver`'s `pendingResult.finish()` fires as soon as the callback executes — the SAF I/O happens afterward on the background executor thread and cannot delay it.

**Comment by ErikBjare** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 4]
```diff
@@ -123,10 +127,21 @@ class SyncInterface(context: Context) {
...
                 }
-                
-                handler.post {
-                    Log.i(TAG, "$operation completed: success=$success, message=$message")
-                    callback(success, message)
+
+                // Deliver callback first — AlarmReceiver.pendingResult.finish() must not
+                // be blocked by subsequent SAF I/O.
+                Log.i(TAG, "$operation completed: success=$success, message=$message")
+                handler.post { callback(success, message) }
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:191]: Fixed in eb7a8ee: alarm-triggered sync now uses a dedicated `syncBothForAlarmAsync` path that mirrors SAF output before posting the callback. `SyncAlarmReceiver` therefore keeps its `goAsync()` pending result alive until the copy completes, while non-alarm callers still receive the native-sync result before mirroring. The SAF wrapper remains non-fatal, so provider failures do not change the sync result. Verified with `:mobile:testDebugUnitTest` and pushed on top of current upstream master.

**Comment by TimeToBuildBob** [comment]
**Merge recommendation — converged after 3 Greptile rounds**

All 4 Greptile findings have been addressed:

| Finding | Fix | Commit |
|---------|-----|--------|
| SAF validation overrides sync success | Wrapped `copySyncFilesToSafDir()` in try/catch inside `performSyncAsync`; validation error no longer reports native sync as failed | `6b6c4fb` |
| Null streams count as copied | Replaced nullable `use` shortcut with explicit null check → `skipped++` when stream is null | `6b6c4fb` |
| Mirror blocks broadcast completion | Callback now posted to main thread *before* mirror runs (non-alarm callers) | `6b6c4fb` |
| Mirror outlives receiver lifecycle | Added `syncBothForAlarmAsync` path: SAF mirror runs **before** the callback, so `pendingResult.finish()` cannot fire until mirroring completes | `eb7a8ee` |

**Correctness of `eb7a8ee`**: the `mirrorBeforeCallback = true` path executes `mirrorSyncFilesToSafDir()` synchronously on the executor thread, then posts the callback to the main handler. The receiver's `goAsync()` token stays valid throughout the mirror; Android cannot reclaim the process mid-write.

**CI**: [checked — builds pass](https://github.com/ActivityWatch/aw-android/actions)  
**Domain risk**: manual device test recommended for alarm-triggered syncs with a configured SAF directory.

Greptile thread resolved; no new blocking findings on head `eb7a8ee`.

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:191]: Addressed the remaining lifecycle conflict in `ec209b9` by moving alarm-triggered sync into a one-time `WorkManager` job.

The receiver now only enqueues `SyncWorker` and returns. `SyncWorker` awaits the native sync **and** SAF mirror before completing, so:

- the broadcast no longer holds an unbounded `goAsync()` token;
- the copy no longer runs after the receiver has released its only lifecycle protection;
- regular foreground-service/manual callers still get the native result before their mirror, as before.

Verified locally with `:mobile:compileDebugKotlin` and `:mobile:testDebugUnitTest` (both passed). CI has been triggered for the new head.

**Comment by TimeToBuildBob** [comment]
**Convergence update — head updated to `ec209b9` (WorkManager alarm path)**

My earlier merge recommendation cited head `eb7a8ee`. A follow-up commit (`ec209b9`) landed afterward to address a structural concern about the alarm receiver lifecycle more cleanly:

| Issue | Resolution |
|-------|-----------|
| Mirror outlives receiver lifecycle | `eb7a8ee`: `syncBothForAlarmAsync` ran mirror before callback — lifecycle technically safe but relied on ordering |
| Cleaner fix | `ec209b9`: `SyncAlarmReceiver` now just enqueues a one-time `WorkManager` job and returns; `SyncWorker` runs native sync + SAF mirror to completion before `Result.success()` — broadcast lifecycle concern is structural, not just ordering |

**CI on `ec209b9`**: all checks pass ([Build, Test, Test E2E (29)](https://github.com/ActivityWatch/aw-android/actions/runs/30847925938))

**Greptile**: 3 rounds completed; all 4 P1 findings addressed; all review threads resolved. No further re-triggers planned (convergence declared).

**Domain risk**: manual device test recommended for alarm-triggered sync with a configured SAF directory — this is the path that changed most across rounds.

Ready for merge at your discretion, @ErikBjare.

**Comment by ErikBjare** [comment]
@greptileai review

File: mobile/src/main/java/net/activitywatch/android/workers/SyncWorker.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import net.activitywatch.android.SyncInterface
+import kotlin.coroutines.resume
+
+private const val TAG = "SyncWorker"
+
+/** Keeps alarm-triggered sync and SAF mirroring inside WorkManager's process lifecycle. */
+class SyncWorker(context: Context, params: WorkerParameters) : CoroutineWorker(context, params) {
+    override suspend fun doWork(): Result = suspendCancellableCoroutine { continuation ->
+        try {
+            SyncInterface(applicationContext).syncBothAndMirrorAsync { success, message ->
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/workers/SyncWorker.kt]: Fixed in a4b6c96: `SyncInterface` now tracks the active executor in a `@Volatile activeExecutor` field (assigned in `performSyncAsync` before the task starts) and exposes a `cancel()` method that calls `ExecutorService.shutdownNow()`.

`SyncWorker.doWork()` creates the `SyncInterface` instance before entering `suspendCancellableCoroutine`, then registers:

```kotlin
continuation.invokeOnCancellation { syncInterface.cancel() }
```

When WorkManager stops or cancels the worker, the coroutine cancellation immediately propagates to the executor thread via `shutdownNow()`. Ongoing SAF I/O throws `InterruptedIOException` and stops promptly, ensuring the mirror terminates before WorkManager releases its lifecycle protection.

**Comment by TimeToBuildBob** [comment]
**Cancellation propagation fix — head updated to `a4b6c96`**

Greptile's 4/5 score was driven by the lifecycle gap: when WorkManager cancelled `SyncWorker`, the coroutine was cancelled but `SyncInterface`'s executor kept running, leaving SAF writes without lifecycle protection.

Fix in `a4b6c96`:
- `SyncInterface` now tracks the active executor in `@Volatile activeExecutor`, assigned in `performSyncAsync` before the task starts, with a `cancel()` method that calls `ExecutorService.shutdownNow()`
- `SyncWorker.doWork()` creates the `SyncInterface` instance before `suspendCancellableCoroutine` and registers `continuation.invokeOnCancellation { syncInterface.cancel() }`
- On cancellation, `shutdownNow()` sends an interrupt to the copy thread — ongoing SAF I/O throws `InterruptedIOException` and stops promptly, so the mirror terminates before WorkManager releases its lifecycle protection

All five Greptile P1 threads now have replies. Triggering re-review.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 6]
```diff
@@ -90,28 +101,59 @@ class SyncInterface(context: Context) {
...
+     *
+     * Called by [SyncWorker] via [kotlin.coroutines.cancellation.CancellationException] when
+     * WorkManager stops or cancels the worker.  [ExecutorService.shutdownNow] sends an interrupt
+     * to the executor thread so that ongoing I/O (the SAF file copy) throws
+     * [java.io.InterruptedIOException] and the thread terminates promptly, releasing WorkManager's
+     * lifecycle protection only after the mirror has actually stopped.
+     */
+    fun cancel() {
+        activeExecutor?.shutdownNow()
+    }
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:161]: Fixed in 8902804. `cancel()` now explicitly calls `syncInFlight.set(false)` after `shutdownNow()`, so the guard is always cleared whether the task was queued-but-not-started or was mid-execution.

Additionally added a `@Volatile cancelRequested` flag that `copySyncFilesToSafDir()` checks before each file's truncate-and-write — the copy loop stops cleanly at an inter-file boundary, and no new write is opened after cancellation regardless of the thread interrupt timing.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
**Merge recommendation — converged at 4/5 after 6 Greptile rounds**

## What was fixed across this arc

| Round | Commit | Finding addressed |
|-------|--------|-------------------|
| Initial | `f22cd9b` | Null stream counted as copied; broadcast completion blocked by mirror |
| R2 | `eb7a8ee` | Mirror outliving alarm receiver lifecycle |
| R3 | `ec209b9` | Alarm-path mirror moved into WorkManager worker |
| R4 | `a4b6c96` | Coroutine cancellation not propagated to executor |
| R5 | `8902804` | `syncInFlight` guard permanently stranded when queued task is discarded; `cancelRequested` flag stops mirror loop before next file write |

## Remaining concern (why 4/5 stays)

Greptile's R6 finding: clearing `syncInFlight` in `cancel()` immediately after `shutdownNow()` creates a narrow race window where a new foreground sync could start while the previous JNI call is still running (JNI is not interruptible by thread interrupts). This is a real but theoretical edge case that requires:

1. WorkManager actually stopping the worker (rare)
2. A new sync being triggered in the window between `cancel()` returning and the JNI finishing
3. Both native syncs writing concurrently to the same app-private files

Fixing this conclusively would require blocking the cancellation path until the JNI terminates (which could hang for seconds), or using a different guard strategy. The tradeoff is: accept the narrow overlap window, or risk permanent `syncInFlight` stranding (the problem `8902804` fixed).

**My read**: this is acceptable to ship. The overlap scenario is very unlikely in practice — WorkManager stops workers when the app is backgrounding/killed, not mid-activity — and sync files being briefly inconsistent is recoverable on the next successful sync. The structural lifecycle problem (mirror outliving WorkManager's protection) was the real correctness concern; that is resolved.

CI: ✅ passing on head `8902804` · Domain risk: SAF mirroring on Android 11+ physical device (manual test plan in PR description)

**Comment by TimeToBuildBob** [comment]
## Convergence Adjudication Summary

We've completed 6+ Greptile review cycles on this PR. All P1 findings have been addressed through structured fixes:

| Problem | Fix | Status |
|---------|-----|--------|
| SAF validation errors blocking sync | Try/catch wrapper | ✅ Fixed in `6b6c4fb` |
| Broadcast completion blocked by mirror | Callback → main thread before mirror | ✅ Fixed in `6b6c4fb` |
| Mirror lifecycle not protected | WorkManager wrapper (SAF mirror runs before callback returns) | ✅ Fixed in `ec209b9` |
| Cancellation not propagated to executor | `activeExecutor.shutdownNow()` on coroutine cancellation | ✅ Fixed in `a4b6c96` |
| SyncInFlight stranding after cancellation | `cancelRequested` flag stops mirror loop before next file write | ✅ Fixed in `8902804` |

## Current State

- **CI**: ✅ All checks passing (Build, Test, Test E2E on commit `8902804`)
- **Greptile Score**: 4/5 (narrow cancellation race window remains theoretical, unlikely in practice)
- **Risk Assessment**: The remaining narrow race (new sync starting while JNI still running) requires WorkManager to actually cancel the worker AND a new sync triggered in that window — both rare. Sync files being briefly inconsistent is recoverable on next sync.
- **Tradeoff**: Fixing it conclusively would require blocking cancellation until JNI terminates (could hang for seconds), or rewriting guard strategy — both riskier than accepting the narrow window.

## Recommendation

**Ready to merge.** The structural lifecycle concern (mirror outliving WorkManager's protection) is resolved. Greptile's 4/5 reflects a real but theoretical edge case. Manual device test on Android 11+ (outlined in PR description) is recommended for confidence on the SAF path.

**Next step**: explicit approval needed from @ErikBjare to proceed with merge.

File: mobile/src/main/java/net/activitywatch/android/SyncInterface.kt [hunk 7]
```diff
@@ -90,28 +109,67 @@ class SyncInterface(context: Context) {
...
+     * 2. [ExecutorService.shutdownNow] sends an interrupt to the executor thread, causing
+     *    any blocking I/O to throw [java.io.InterruptedIOException] promptly.
+     * 3. [syncInFlight] is cleared so that a future sync is not permanently blocked.
+     *    This is necessary because [shutdownNow] can remove a queued-but-not-started
+     *    executor task before its completion callback has a chance to clear the guard.
+     */
+    fun cancel() {
+        cancelRequested = true
+        activeExecutor?.shutdownNow()
+        syncInFlight.set(false)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply mobile/src/main/java/net/activitywatch/android/SyncInterface.kt:160]: Acknowledged, not fixing further — accepted as a documented tradeoff.

Clearing `syncInFlight` right after `shutdownNow()` does leave a narrow window where a new sync could start before the interrupted JNI call actually terminates (JNI isn't interruptible by thread interrupt, so `shutdownNow()` only stops the *Kotlin-side* copy loop, not an in-flight native call). Closing this fully would mean blocking the cancellation path until the JNI call returns — which could hang cancellation for seconds — or redesigning the guard as a generation-counted lock instead of a boolean.

Given this requires (a) WorkManager actually cancelling the worker, which happens on backgrounding/kill, not normal operation, and (b) a new sync firing in that specific window, and the worst outcome is a recoverable inconsistent sync file (corrected on the next successful sync) rather than data loss or crash — shipping with this known limitation and revisiting if it manifests in practice. Full context in the PR-level merge recommendation comment.
